### PR TITLE
Add SQLTest

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add https://github.com/elliotchance/sqltest to CI to report the SQL 2016 conformance.

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 8e3bd222f00324667a1d41fb62f50adc91045b6c with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52293/8e3bd222f00324667a1d41fb62f50adc91045b6c/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: docker/images.json [hunk 1]
```diff
@@ -155,11 +155,14 @@
...
         "name": "clickhouse/docs-builder",
-        "dependent": [
-        ]
+        "dependent": []
     },
     "docker/test/sqllogic": {
         "name": "clickhouse/sqllogic-test",
         "dependent": []
+    },
+    "docker/test/sqltest": {
```
Comment (Felixoid) [inline docker/images.json:164]: you must add it to the dependents of `clickhouse/test-base`

File: tests/ci/tests/docker_images.json [hunk 2]
```diff
@@ -153,5 +153,9 @@
     "docker/test/sqllogic": {
         "name": "clickhouse/sqllogic-test",
         "dependent": []
+    },
+    "docker/test/sqltest": {
```
Comment (Felixoid) [inline tests/ci/tests/docker_images.json:157]: here too, please

File: docker/test/sqltest/run.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+# shellcheck disable=SC2015
+
+set -x
+set -e
+set -u
+set -o pipefail
+
+BINARY_TO_DOWNLOAD=${BINARY_TO_DOWNLOAD:="clang-16_debug_none_unsplitted_disable_False_binary"}
```
Comment (Felixoid) [inline docker/test/sqltest/run.sh:9]: Please, don't use the worst practices from performance tests. Binaries should be downloaded in advance and mounted into the container by a script from `tests/ci`

File: tests/ci/sqltest.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        os.makedirs(workspace_path)
+
+    run_command = get_run_command(
+        pr_info.number, pr_info.sha, build_url, workspace_path, docker_image
+    )
+    logging.info("Going to run %s", run_command)
+
+    run_log_path = os.path.join(temp_path, "run.log")
+    with open(run_log_path, "w", encoding="utf-8") as log:
+        with subprocess.Popen(
```
Comment (Felixoid) [inline tests/ci/sqltest.py:96]: `TeePopen` looks like a better class here

File: tests/ci/sqltest.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    subprocess.check_call(f"sudo chown -R ubuntu:ubuntu {temp_path}", shell=True)
+
+    check_name_lower = (
+        check_name.lower().replace("(", "").replace(")", "").replace(" ", "")
+    )
+    s3_prefix = f"{pr_info.number}/{pr_info.sha}/sqltest_{check_name_lower}/"
+    paths = {
+        "run.log": run_log_path,
+        "main.log": os.path.join(workspace_path, "main.log"),
```
Comment (Felixoid) [inline tests/ci/sqltest.py:113]: just a side note, py3 has `pathlib.Path` class. It's much nicer in usage than the old `os.path` things

File: tests/ci/sqltest.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        "server.log.zst": os.path.join(workspace_path, "server.log.zst"),
+        "server.err.log.zst": os.path.join(workspace_path, "server.err.log.zst"),
+        "report.html": os.path.join(workspace_path, "report.html"),
+        "test.log": os.path.join(workspace_path, "test.log"),
+    }
+
+    s3_helper = S3Helper()
+    for f in paths:
+        try:
+            paths[f] = s3_helper.upload_test_report_to_s3(paths[f], s3_prefix + f)
```
Comment (Felixoid) [inline tests/ci/sqltest.py:123]: by adding logs to `TestReport` you can skip this part

File: tests/ci/sqltest.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            logging.info("Exception uploading file %s text %s", f, ex)
+            paths[f] = ""
+
+    report_url = GITHUB_RUN_URL
+    if paths["report.html"]:
+        report_url = paths["report.html"]
+
+    status = "success"
+    description = "See the report"
+    test_result = TestResult(description, "OK")
```
Comment (Felixoid) [inline tests/ci/sqltest.py:134]: add logs and report files here

**Review by Felixoid — CHANGES_REQUESTED** [review]

File: docker/test/sqltest/run.sh (hunk 3 above)
Comment (alexey-milovidov) [reply docker/test/sqltest/run.sh:9]: Ok. I use copy-paste and ChatGPT to write Python. And when I'm using copy-paste, I'm trying to reuse the shortest code, so I have less amount to read.

**Comment by alexey-milovidov** [comment]
@Felixoid currently it does not work due to

```
FAIL: test_get_changed_docker_images (docker_test.TestDockerImageCheck)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/ci/docker_test.py", line 99, in test_get_changed_docker_images
    self.assertEqual(images, expected)
AssertionError: Lists differ: [Dock[675 chars]est/sqltest,repo=clickhouse/sqltest,parent=cli[410 chars]est)] != [Dock[675 chars]est/stateful,repo=clickhouse/stateful-test,par[319 chars]est)]

First differing element 7:
Docke[20 chars]est/sqltest,repo=clickhouse/sqltest,parent=cli[13 chars]base)
Docke[20 chars]est/stateful,repo=clickhouse/stateful-test,par[25 chars]test)

First list contains 1 additional elements.
First extra element 11:
DockerImage(path=docker/test/unit,repo=clickhouse/unit-test,parent=clickhouse/stateless-test)

  [DockerImage(path=docker/docs/builder,repo=clickhouse/docs-builder,parent=None),
   DockerImage(path=docker/test/base,repo=clickhouse/test-base,parent=None),
   DockerImage(path=docker/docs/check,repo=clickhouse/docs-check,parent=clickhouse/docs-builder),
   DockerImage(path=docker/docs/release,repo=clickhouse/docs-release,parent=clickhouse/docs-builder),
   DockerImage(path=docker/test/fuzzer,repo=clickhouse/fuzzer,parent=clickhouse/test-base),
   DockerImage(path=docker/test/integration/base,repo=clickhouse/integration-test,parent=clickhouse/test-base),
   DockerImage(path=docker/test/keeper-jepsen,repo=clickhouse/keeper-jepsen-test,parent=clickhouse/test-base),
-  DockerImage(path=docker/test/sqltest,repo=clickhouse/sqltest,parent=clickhouse/test-base),
   DockerImage(path=docker/test/stateful,repo=clickhouse/stateful-test,parent=clickhouse/stateless-test),
   DockerImage(path=docker/test/stateless,repo=clickhouse/stateless-test,parent=clickhouse/test-base),
   DockerImage(path=docker/test/stress,repo=clickhouse/stress-test,parent=clickhouse/stateful-test),
   DockerImage(path=docker/test/unit,repo=clickhouse/unit-test,parent=clickhouse/stateless-test)]
```

**Comment by alexey-milovidov** [comment]
What is the difference between `images.json` and `docker_images.json`?
They look almost identical, but not quite. And JSON does not support comments, unfortunately.

**Comment by alexey-milovidov** [comment]
I found something strange, but didn't understand it: https://github.com/ClickHouse/ClickHouse/pull/33751

**Comment by alexey-milovidov** [comment]
@Felixoid Now it is complete ..F.........

> First differing element 7:
Docke[21 chars]st/sqltest,repo=clickhouse/sqltest,parent=clickhouse/test-base)
Docke[21 chars]st/sqltest,repo=clickhouse/sqltest,parent=clickhouse/test-base)

**Comment by Felixoid** [comment]
You shouldn't have used `True` for [only_amd64](https://github.com/ClickHouse/ClickHouse/blob/master/tests/ci/docker_images_check.py#L51)

File: .github/workflows/pull_request.yml [hunk 8]
```diff
@@ -5182,3 +5182,39 @@ jobs:
...
+  SQLTest:
+    needs: [BuilderDebRelease]
+    runs-on: [self-hosted, fuzzer-unit-tester]
+    steps:
+      - name: Set envs
+        run: |
+          cat >> "$GITHUB_ENV" << 'EOF'
+          TEMP_PATH=${{runner.temp}}/sqltest
+          REPORTS_PATH=${{runner.temp}}/reports_dir
+          CHECK_NAME="SQLTest"
```
Comment (Felixoid) [inline .github/workflows/pull_request.yml:5197]: https://github.com/ClickHouse/ClickHouse/actions/runs/5731581942/job/15534600187?pr=52293#step:6:12

You use it literally with quotes

**Comment by alexey-milovidov** [comment]
I thought Azure test was fixed...

File: tests/ci/sqltest.py (hunk 7 above)
Comment (alexey-milovidov) [reply tests/ci/sqltest.py:133]: Ok. I want to make a small working prototype first.
This check has limited use. It should exist; everything else we will address later (or never).

**Pull request merged**